File: runtime/python/phonebook_rt/collections_.py
"""Area 300 — collections.

The ordering promises live here: ENTRIES sorts by key, SORT and SORT_BY are
stable, UNIQUE keeps first occurrences. Those are the rules that make a
map-derived result identical in a language with insertion-ordered dicts and a
language with key-ordered trees.
"""

from __future__ import annotations


def make_list(*items):
    """300-0000001 LIST."""
    return list(items)


def filter_seq(sequence: list, predicate):
    """300-0000002 FILTER — order preserved, predicate applied once per item."""
    return [item for item in sequence if predicate(item)]


def map_seq(sequence: list, transform):
    """300-0000003 MAP."""
    return [transform(item) for item in sequence]


def reduce_seq(sequence: list, combine, initial):
    """300-0000004 REDUCE — left fold, always."""
    accumulator = initial
    for item in sequence:
        accumulator = combine(accumulator, item)
    return accumulator


def sort_seq(sequence: list):
    """300-0000005 SORT — stable; text orders by code point, not locale."""
    return sorted(sequence)


def sort_by(sequence: list, key, descending: bool):
    """300-0000006 SORT_BY — stable in both directions.

    `reverse=True` reverses the comparison, not the result, so ties keep their
    original relative order. Reversing the sorted list instead would silently
    break that, and the tie order is exactly what makes word_freq's output
    reproducible.
    """
    return sorted(sequence, key=key, reverse=descending)


def reverse_seq(sequence: list):
    """300-0000007 REVERSE."""
    return list(reversed(sequence))


def unique(sequence: list):
    """300-0000008 UNIQUE — first-occurrence order."""
    return list(dict.fromkeys(sequence))


def count(sequence: list) -> int:
    """300-0000009 COUNT."""
    return len(sequence)


def take(sequence: list, n: int):
    """300-0000010 TAKE — clamped at both ends, never an error."""
    if n <= 0:
        return []
    return sequence[:n]


def first(sequence: list, fallback):
    """300-0000011 FIRST — total by construction."""
    return sequence[0] if sequence else fallback


def count_occurrences(sequence: list) -> dict:
    """300-0000012 COUNT_OCCURRENCES."""
    tallies: dict = {}
    for item in sequence:
        tallies[item] = tallies.get(item, 0) + 1
    return tallies


def entries(mapping: dict) -> list:
    """300-0000013 ENTRIES — sorted by key, by contract."""
    return [(key, mapping[key]) for key in sorted(mapping)]


def get(mapping: dict, key, fallback):
    """300-0000014 GET — total: a missing key returns the fallback."""
    return mapping.get(key, fallback)


def pair_key(value):
    """300-0000015 PAIR_KEY."""
    return value[0]


def pair_value(value):
    """300-0000016 PAIR_VALUE."""
    return value[1]

File: runtime/rust/phonebook_rt/src/collections_.rs
//! Area 300 — collections.
//!
//! The ordering promises live here. `ENTRIES` sorting by key costs nothing in
//! Rust, because `BTreeMap` already iterates in key order — but the Python side
//! has to sort explicitly, since its dicts iterate in insertion order. Neither
//! backend gets to decide; the contract did.

use std::collections::{BTreeMap, BTreeSet};

/// 300-0000002 FILTER — order preserved, predicate applied once per item.
pub fn filter_seq<T, F>(sequence: &[T], predicate: F) -> Vec<T>
where
    T: Clone,
    F: Fn(&T) -> bool,
{
    sequence
        .iter()
        .filter(|item| predicate(item))
        .cloned()
        .collect()
}

/// 300-0000003 MAP
pub fn map_seq<T, R, F>(sequence: &[T], transform: F) -> Vec<R>
where
    F: Fn(&T) -> R,
{
    sequence.iter().map(transform).collect()
}

/// 300-0000004 REDUCE — left fold, always.
pub fn reduce_seq<T, A, F>(sequence: &[T], combine: F, initial: &A) -> A
where
    A: Clone,
    F: Fn(&A, &T) -> A,
{
    let mut accumulator = initial.clone();
    for item in sequence {
        accumulator = combine(&accumulator, item);
    }
    accumulator
}

/// 300-0000005 SORT — stable; `String` orders by byte, which for UTF-8 is
/// exactly code-point order, so this agrees with Python without special care.
pub fn sort_seq<T: Clone + Ord>(sequence: &[T]) -> Vec<T> {
    let mut out = sequence.to_vec();
    out.sort();
    out
}

/// 300-0000006 SORT_BY — stable in both directions.
///
/// Descending reverses the *comparison*, not the result. Reversing the sorted
/// vector instead would invert the order of equal keys and quietly break the
/// contract's stability promise.
pub fn sort_by<T, K, F>(sequence: &[T], key: F, descending: &bool) -> Vec<T>
where
    T: Clone,
    K: Ord,
    F: Fn(&T) -> K,
{
    let mut out = sequence.to_vec();
    if *descending {
        out.sort_by_key(|item| std::cmp::Reverse(key(item)));
    } else {
        out.sort_by_key(&key);
    }
    out
}

/// 300-0000007 REVERSE
pub fn reverse_seq<T: Clone>(sequence: &[T]) -> Vec<T> {
    sequence.iter().rev().cloned().collect()
}

/// 300-0000008 UNIQUE — first-occurrence order.
pub fn unique<T: Clone + Ord>(sequence: &[T]) -> Vec<T> {
    let mut seen: BTreeSet<T> = BTreeSet::new();
    let mut out: Vec<T> = Vec::new();
    for item in sequence {
        if seen.insert(item.clone()) {
            out.push(item.clone());
        }
    }
    out
}

/// 300-0000009 COUNT
pub fn count<T>(sequence: &[T]) -> i64 {
    sequence.len() as i64
}

/// 300-0000010 TAKE — clamped at both ends, never an error.
pub fn take<T: Clone>(sequence: &[T], n: &i64) -> Vec<T> {
    if *n <= 0 {
        return Vec::new();
    }
    let limit = (*n as usize).min(sequence.len());
    sequence[..limit].to_vec()
}

/// 300-0000011 FIRST — total by construction.
pub fn first<T: Clone>(sequence: &[T], fallback: &T) -> T {
    sequence
        .first()
        .cloned()
        .unwrap_or_else(|| fallback.clone())
}

/// 300-0000012 COUNT_OCCURRENCES
pub fn count_occurrences<T: Clone + Ord>(sequence: &[T]) -> BTreeMap<T, i64> {
    let mut tallies: BTreeMap<T, i64> = BTreeMap::new();
    for item in sequence {
        *tallies.entry(item.clone()).or_insert(0) += 1;
    }
    tallies
}

/// 300-0000013 ENTRIES — sorted by key, by contract.
pub fn entries<K: Clone + Ord, V: Clone>(mapping: &BTreeMap<K, V>) -> Vec<(K, V)> {
    mapping
        .iter()
        .map(|(key, value)| (key.clone(), value.clone()))
        .collect()
}

/// 300-0000014 GET — total: a missing key returns the fallback.
pub fn get<K: Ord, V: Clone>(mapping: &BTreeMap<K, V>, key: &K, fallback: &V) -> V {
    mapping
        .get(key)
        .cloned()
        .unwrap_or_else(|| fallback.clone())
}

/// 300-0000015 PAIR_KEY
pub fn pair_key<K: Clone, V>(value: &(K, V)) -> K {
    value.0.clone()
}

/// 300-0000016 PAIR_VALUE
pub fn pair_value<K, V: Clone>(value: &(K, V)) -> V {
    value.1.clone()
}
